improver/spotdata/spot_extraction.py

```diff
@@ -124,18 +124,31 @@ def extract_diagnostic_data(coordinate_cube, diagnostic_cube):
                 An array of diagnostic values at the grid coordinates found
                 within the coordinate cube.
         """
-        enforce_coordinate_ordering(
-            diagnostic_cube,
-            [
-                diagnostic_cube.coord(axis="x").name(),
-                diagnostic_cube.coord(axis="y").name(),
-            ],
-        )
-        spot_values = diagnostic_cube.data[tuple(coordinate_cube.data.T)]
-        return spot_values
+        yname = diagnostic_cube.coord(axis="y").name()
+        xname = diagnostic_cube.coord(axis="x").name()
+        if (
+            diagnostic_cube.dim_coords[-1].name() != xname
+            and diagnostic_cube.dim_coords[-2].name() != yname
+        ):
+            # Collect all the names, move y and x to the end and reorder the diagnostic cube
+            dims = [dim.name() for dim in diagnostic_cube.dim_coords]
+            dims.remove(xname)
+            dims.remove(yname)
+            dims.extend([yname, xname])
+            new_order = [diagnostic_cube.coord_dims(dim)[0] for dim in dims]
+            diagnostic_cube.transpose(new_order)
+
+        coords = coordinate_cube.data
+        return diagnostic_cube.data[..., coords[1], coords[0]]
```

Comment:
A total nitpick, but you could unpack x-y indices in a tuple assignment for better readability:
```suggestion
        x_indices, y_indices = coordinate_cube.data
        return diagnostic_cube.data[..., y_indices, x_indices]
```